File: LeetCode/LeetCode/Problems/18/Problem1845.cs
using System.Collections.Generic;
using System.Linq;

public class Problem1845
{
    public class SeatManager
    {
        private int _last = 0;
        private readonly SortedSet<int> _freeSeats = new SortedSet<int>();
        public SeatManager(int n)
        {
        }

        public int Reserve()
        {
            if (!_freeSeats.Any())
                return ++_last;
            else
            {
                var num = _freeSeats.Min;
                _freeSeats.Remove(num);
                return num;
            }
        }

        public void Unreserve(int seatNumber)
        {
            if (seatNumber == _last)
                _last--;
            else
                _freeSeats.Add(seatNumber);
        }
    }
}

File: leetcode_rust/src/problems/problem_1845.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

struct SeatManager {
    last: i32,
    free_seats: BinaryHeap<Reverse<i32>>,
}

impl SeatManager {
    fn new(_: i32) -> Self {
        SeatManager {
            last: 0,
            free_seats: BinaryHeap::new(),
        }
    }

    fn reserve(&mut self) -> i32 {
        if self.free_seats.is_empty() {
            self.last += 1;
            self.last
        } else {
            self.free_seats.pop().unwrap().0
        }
    }

    fn unreserve(&mut self, seat_number: i32) {
        if seat_number == self.last {
            self.last -= 1;
        } else {
            self.free_seats.push(Reverse(seat_number));
        }
    }
}
